### PR TITLE
Base: LayerRAMSubSet algorithm

### PR DESCRIPTION
**Pull request by martinfalk**
* heavily based on VolumeRAMSubSet

### CODE REVIEW

File: modules/base/algorithm/image/layerramsubset.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+namespace inviwo {
+
+std::shared_ptr<LayerRAM> LayerRAMSubSet::apply(const LayerRepresentation *in, ivec2 offset,
+                                                size2_t extent, bool clampBorderOutsideImage,
+                                                const DataFormatBase *dstFormat) {
+    if (dstFormat) {
+#include <warn/push>
+#include <warn/ignore/switch-enum>
+        switch (dstFormat->getId()) {
+            case DataFormatId::Vec4UInt8: {
```
Comment (petersteneteg) [inline modules/base/algorithm/image/layerramsubset.cpp]: Only vec4uint8 dest supported?

File: modules/base/algorithm/image/layerramsubset.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                detail::LayerRAMSubSetDispatcher disp;
+                return dispatching::dispatch<std::shared_ptr<LayerRAM>, dispatching::filter::All>(
+                    dstFormat->getId(), disp, in, offset, extent, clampBorderOutsideImage);
+            }
+        }
+#include <warn/pop>
+    } else {
+        detail::LayerRAMSubSetDispatcher disp;
+        return dispatching::dispatch<std::shared_ptr<LayerRAM>, dispatching::filter::All>(
+            dstFormat->getId(), disp, in, offset, extent, clampBorderOutsideImage);
```
Comment (petersteneteg) [inline modules/base/algorithm/image/layerramsubset.cpp]: DstFormat == null?

File: modules/base/algorithm/image/layerramsubset.cpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+std::shared_ptr<LayerRAM> LayerRAMSubSet::apply(const LayerRepresentation *in, ivec2 offset,
+                                                size2_t extent, bool clampBorderOutsideImage,
+                                                const DataFormatBase *dstFormat) {
+    if (dstFormat) {
+#include <warn/push>
+#include <warn/ignore/switch-enum>
+        switch (dstFormat->getId()) {
+            case DataFormatId::Vec4UInt8: {
+                detail::LayerRAMSubSetConvertDispatcher<DataVec4UInt8::type> disp;
+                return dispatching::dispatch<std::shared_ptr<LayerRAM>, dispatching::filter::All>(
```
Comment (petersteneteg) [inline modules/base/algorithm/image/layerramsubset.cpp]: Why not dispatch on the layer directly?

File: modules/base/algorithm/image/layerramsubset.cpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+                    dstFormat->getId(), disp, in, offset, extent, clampBorderOutsideImage);
+            }
+            default: {
+                detail::LayerRAMSubSetDispatcher disp;
+                return dispatching::dispatch<std::shared_ptr<LayerRAM>, dispatching::filter::All>(
+                    dstFormat->getId(), disp, in, offset, extent, clampBorderOutsideImage);
```
Comment (petersteneteg) [inline modules/base/algorithm/image/layerramsubset.cpp]: Should the dispatch not be in the source format?

File: modules/base/algorithm/image/layerramsubset.h [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        // clear entire layer as only parts will be copied
+        std::fill(dst, dst + dstDim.x * dstDim.y, Tdst(0));
+    }
+    // memcpy each row to form sub layer
+#pragma omp parallel for
+    for (int j = 0; j < copyExtent.y; j++) {
+        size_t srcPos = (j + srcOffset.y) * srcDim.x + srcOffset.x;
+        size_t dstPos = (j + dstOffset.y) * dstDim.x + dstOffset.x;
+        for (int i = 0; i < copyExtent.x; i++) {
+            dst[dstPos + i] = util::glm_convert_normalized<Tdst, T>(src[srcPos + i]);
```
Comment (petersteneteg) [inline modules/base/algorithm/image/layerramsubset.h]: Could do this with a second dispatch...

**Review by petersteneteg — CHANGES_REQUESTED** [review]

File: modules/base/algorithm/image/layerramsubset.cpp (hunk 1 above)
Comment (martinfalk) [reply modules/base/algorithm/image/layerramsubset.cpp]: regarding conversion as a target format (as per comment). There was no need to implement a subset + format conversion for other formats besides 8bit rgba.

File: modules/base/algorithm/image/layerramsubset.cpp (hunk 3 above)
Comment (martinfalk) [reply modules/base/algorithm/image/layerramsubset.cpp]: This code is approx. 2 years old and was based on `VolumeRAMSubset`, will fix this.

File: modules/base/algorithm/image/layerramsubset.cpp (hunk 4 above)
Comment (martinfalk) [reply modules/base/algorithm/image/layerramsubset.cpp]: hm, might work in connection with the second dispatch.

**Comment by martinfalk** [comment]
Revised PR with respect to requested changes.

**Review by petersteneteg — APPROVED** [review]

**Pull request merged**